xwiki-platform-core/xwiki-platform-rendering/xwiki-platform-rendering-macros/xwiki-platform-rendering-macro-code/src/main/java/org/xwiki/rendering/macro/code/CodeMacroLayout.java

```diff
@@ -0,0 +1,74 @@
+/*
+ * See the NOTICE file distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as
+ * published by the Free Software Foundation; either version 2.1 of
+ * the License, or (at your option) any later version.
+ *
+ * This software is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this software; if not, write to the Free
+ * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+ * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+ */
+package org.xwiki.rendering.macro.code;
+
+import org.xwiki.rendering.code.layout.CodeLayoutHandler;
+
+/**
+ * Values allowed in the <code>layout</code> parameter of the code macro.
+ * 
+ * @version $Id$
+ * @since 11.3
+ */
+public enum CodeMacroLayout
+{
+    /**
+     * Output the blocks as rendered by the parser.
+     */
+    PLAIN(Constants.PLAIN_HINT),
+    /**
+     * Display line numbers beside the rendered code.
+     */
+    LINE_NUM(Constants.LINE_NUM_HINT);
+
+    private String hint;
+
+    CodeMacroLayout(String hint)
+    {
+        this.hint = hint;
+    }
+
+    /**
+     * @return the hint value to request the proper layout handler via DI.
+     */
+    public final String getHint()
+    {
+        return hint;
+    }
+
+    /**
+     * Convenience class holding hint values for {@link CodeLayoutHandler} components. Links the outer enum class to the
+     * components classes as it is not possible to reference enums in annotations (and avoids using plain strings).
+     * 
+     * @version $Id$
+     */
+    public static class Constants
```

Comment:
I don't understand why you introduce a new inner class for that, it looks a bit overly complicated just for 2 strings...

Reply:
A third class is [necessary](https://stackoverflow.com/questions/13253624/how-to-supply-enum-value-to-an-annotation-from-a-constant-in-java) to be able to reference the constants from the enum definition (try removing it). If your question is more "why I used constants", well, in general, I avoid copy-pasting String literals, especially when contributing to a project I'm not familiar with.

If you're sure we're better off without that piece of code I'll remove it.

Reply:
> If your question is more "why I used constants"

no it wasn't, you perfectly right to use constants :) 

My bad, indeed in that case it's difficult to do differently. Now I'm not even sure this enum is needed: you could have two static Strings in CodeMacroParameters and rely only on them... It's really a style matter I guess.